src/abstractions/handler-base.ts

```diff
@@ -0,0 +1,216 @@
+import * as React from "react";
+import * as PropTypes from "prop-types";
+
+import { EventSource } from "../contracts";
+import { CAN_I_USE_WINDOW_LISTENERS, ESCAPE_KEYCODE } from "../helpers";
+
+export interface HandlerBaseProps {
+    defaultOpen?: boolean;
+    open?: boolean;
+    onOpen?: (source: EventSource) => void;
+    onClose?: (source: EventSource) => void;
+    onToggle?: (isOpened: boolean, source: EventSource) => void;
+    toggleOnHeaderClick?: boolean;
+    closeOnOutsideClick?: boolean;
+    closeOnSectionClick?: boolean;
+    closeOnEscapeClick?: boolean;
+}
+
+export interface HandlerBaseState {
+    open: boolean;
+}
+
+export interface BaseHandlerChildContext {
+    dropdownOpen: boolean;
+    dropdownOnHeaderClickCallback: Function;
+    dropdownOnSectionClickCallback: Function;
+}
+
+export abstract class HandlerBase<
+    TProps extends HandlerBaseProps = HandlerBaseProps,
+    TState extends HandlerBaseState = HandlerBaseState
+> extends React.Component<TProps, TState> {
+    constructor(props: TProps) {
+        super(props);
+        if (CAN_I_USE_WINDOW_LISTENERS) {
+            window.addEventListener("click", this.onOutsideClick.bind(this));
+            // TODO: Move keyup to listener to handler container

```

Comment:
New or old TODO?

Reply:
Old one. Not sure if I should leave that.